### PR TITLE
added new webhook constructors

### PR DESCRIPTION
**Pull request by DJOetzi**
one adds the ability to directly use the webhook URL provided by discord, and the other one takes the webhook ID and webhook URL

### CODE REVIEW

File: include/dpp/webhook.h [hunk 1]
```diff
@@ -56,6 +56,24 @@ class DPP_EXPORT webhook : public managed {
 	 */
 	webhook();
 
+	/**
+     * @brief Construct a new webhook object using the Webhook URL provided by Discord
+    */
```
Comment (braindigitalis) [inline include/dpp/webhook.h:61]: need to document the url parameter:
`@param webhook_url a fully qualified web address of an existing webhook`

File: include/dpp/webhook.h [hunk 2]
```diff
@@ -56,6 +56,24 @@ class DPP_EXPORT webhook : public managed {
...
+	webhook(std::string webhook_url)
+	{
+		token = webhook_url.substr(webhook_url.find_last_of("/") + 1); //find token
+		id = std::stoull(webhook_url.substr(sizeof("https://discord.com/api/webhooks/") - 1, 18)); //18 is the amount of characters a snowflake possesses
+	}
+
+	/**
+     * @brief Construct a new webhook object using the webhook ID and the webhook token
```
Comment (braindigitalis) [inline include/dpp/webhook.h:69]: same here, must document the parameters

**Review by braindigitalis — CHANGES_REQUESTED** [review]

File: include/dpp/webhook.h [hunk 3]
```diff
@@ -56,6 +56,24 @@ class DPP_EXPORT webhook : public managed {
 	 */
 	webhook();
 
+	/**
+     * @brief Construct a new webhook object using the Webhook URL provided by Discord
+    */
+	webhook(std::string webhook_url)
+	{
+		token = webhook_url.substr(webhook_url.find_last_of("/") + 1); //find token
+		id = std::stoull(webhook_url.substr(sizeof("https://discord.com/api/webhooks/") - 1, 18)); //18 is the amount of characters a snowflake possesses
```
Comment (es183923) [inline include/dpp/webhook.h:65]: A snowflake doesn't necessarily be 18 characters. There are 17 digit ids, and soon there will be 19 digit ids.

Comment (DJOetzi) [reply include/dpp/webhook.h:65]: Noted that, thanks for pointing it out!

Comment (braindigitalis) [reply include/dpp/webhook.h:65]: also bear in mind std::stoull can throw... if you don't want to bubble that exception to the user you should catch it, and probably re throw it as `dpp::logic_exception`

File: include/dpp/webhook.h [hunk 4]
```diff
@@ -56,6 +56,24 @@ class DPP_EXPORT webhook : public managed {
...
+	}
+
+	/**
+     * @brief Construct a new webhook object using the webhook ID and the webhook token
+    */
+	webhook(snowflake webhook_id, std::string webhook_token)
```
Comment (braindigitalis) [inline include/dpp/webhook.h:71]: constructor body should be in webhook.cpp not .h

**Review by braindigitalis — CHANGES_REQUESTED** [review]

File: include/dpp/webhook.h (hunk 3 above)
Comment (braindigitalis) [reply include/dpp/webhook.h:65]: also ctor body belongs in the cpp or you can end up with it being defined more than once causing a compile error

File: src/dpp/webhook.cpp [hunk 5]
```diff
@@ -33,6 +33,25 @@ webhook::webhook() : managed(), type(w_incoming), guild_id(0), channel_id(0), us
 {
 }
 
+webhook::webhook(std::string webhook_url)
+{
+	token = webhook_url.substr(webhook_url.find_last_of("/") + 1);
+	try
+	{
+		id = std::stoull(webhook_url.substr(sizeof("https://discord.com/api/webhooks/") - 1, webhook_url.find_last_of("/")));
```
Comment (braindigitalis) [inline src/dpp/webhook.cpp:41]: this isn't doing what you think it is, sizeof() on a std::string will return the size of the object not the string length.
use webhook_url.length(), or for a const char*, strlen.

File: src/dpp/webhook.cpp [hunk 6]
```diff
@@ -33,6 +33,25 @@ webhook::webhook() : managed(), type(w_incoming), guild_id(0), channel_id(0), us
...
+	}
+	catch (std::exception& e)
+	{
+		throw dpp::logic_exception(std::string("Your webhook's URL failed to parse! ID invalid.\n") + e.what());
+	}
+}
+
+webhook::webhook(snowflake webhook_id, std::string webhook_token)
```
Comment (braindigitalis) [inline src/dpp/webhook.cpp:49]: string parameters should be `const std::string&`

Comment (DJOetzi) [reply src/dpp/webhook.cpp:49]: the ID can be made a constant too, right?

Comment (Axyte) [reply src/dpp/webhook.cpp:49]: Yes.

File: src/dpp/webhook.cpp [hunk 7]
```diff
@@ -33,6 +33,25 @@ webhook::webhook() : managed(), type(w_incoming), guild_id(0), channel_id(0), us
...
+{
+	token = webhook_token;
+	id = webhook_id;
```
Comment (Axyte) [inline src/dpp/webhook.cpp:52]: Consider using initialiser lists, if assigning is all you do in a constructor.

File: include/dpp/webhook.h [hunk 8]
```diff
@@ -56,6 +56,21 @@ class DPP_EXPORT webhook : public managed {
 	 */
 	webhook();
```
Comment (Axyte) [inline include/dpp/webhook.h:57]: Maybe mark it as default?

File: src/dpp/webhook.cpp (hunk 7 above)
Comment (Axyte) [reply src/dpp/webhook.cpp:52]: I think you also need to initialise the other data members, as in the original constructor?

**Pull request merged**